File: backend/src/main/java/com/corvuscode/profile/service/ProfileService.java
package com.corvuscode.profile.service;

import com.corvuscode.profile.dto.ChangePasswordRequest;
import com.corvuscode.profile.dto.ProfileResponse;
import com.corvuscode.profile.dto.PublicProfileResponse;
import com.corvuscode.profile.dto.UpdateProfileRequest;

public interface ProfileService {

    ProfileResponse getMyProfile(String email);

    ProfileResponse updateMyProfile(
            String email,
            UpdateProfileRequest request
    );

    void changePassword(
            String email,
            ChangePasswordRequest request
    );

    PublicProfileResponse getPublicProfile(Long userId);

    ProfileResponse getProfileByUserId(Long userId);
}

File: frontend/src/services/profile.service.ts
import api from "./api";

export interface ProfileData {
  id: number;
  fullName: string;
  email: string;
  bio?: string;
  avatarUrl?: string;
  country?: string;
  githubUrl?: string;
  linkedinUrl?: string;
  websiteUrl?: string;
  rating: number;
  totalSolved: number;
  enabled: boolean;
  createdAt: string;
  updatedAt: string;
  roles: string[];
}

export interface PublicProfileData {
  id: number;
  fullName: string;
  bio?: string;
  avatarUrl?: string;
  country?: string;
  githubUrl?: string;
  linkedinUrl?: string;
  websiteUrl?: string;
  rating: number;
  totalSolved: number;
  createdAt: string;
}

export interface UpdateProfileRequest {
  fullName: string;
  bio?: string;
  country?: string;
  avatarUrl?: string;
  githubUrl?: string;
  linkedinUrl?: string;
  websiteUrl?: string;
}

export interface ChangePasswordRequest {
  currentPassword: string;
  newPassword: string;
}

export const profileService = {
  getMyProfile() {
    return api.get<ProfileData>("/profile/me");
  },

  updateMyProfile(data: UpdateProfileRequest) {
    return api.put<ProfileData>("/profile/me", data);
  },

  changePassword(data: ChangePasswordRequest) {
    return api.put<void>("/profile/me/password", data);
  },

  getPublicProfile(userId: number) {
    return api.get<PublicProfileData>(`/profile/${userId}`);
  },
};
